stock_move_actual_date/models/stock_valuation_layer.py

```diff
@@ -0,0 +1,37 @@
+# Copyright 2025 Quartile Limited
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class StockValuationLayer(models.Model):
+    _inherit = "stock.valuation.layer"
+
+    actual_date = fields.Date(
+        compute="_compute_actual_date",
+        store=True,
+        help=(
+            "The actual date is determined as follows:\n"
+            "- If a posted journal entry exists, its date is used.\n"
+            "- If there is no journal entry, the stock move's actual date is used.\n"
+            "- Otherwise, the record's creation date (timezone-aware) is used."
+        ),
+    )
+
+    @api.depends(
+        "create_date",
+        "account_move_id.state",
+        "account_move_id.date",
+        "stock_move_id.actual_date",
+        "stock_move_id.state",
+    )
+    def _compute_actual_date(self):
+        for rec in self:
+            account_move = rec.account_move_id
+            if not account_move:
+                rec.actual_date = rec.stock_move_id.actual_date
+                continue
+            if account_move and account_move.state == "posted":
+                rec.actual_date = account_move.date
+                continue
```

Comment:
To ensure that `actual_date` is always set even when there is neither an account move nor a stock move associated (e.g., records from valuation method change from manual to real time for zero-cost products).

```suggestion
            if account_move and account_move.state == "posted":
                rec.actual_date = account_move.date
                continue
            if rec.stock_move_id.actual_date:
                rec.actual_date = rec.stock_move_id.actual_date
                continue
```